rule/method/src/main/java/io/opentracing/contrib/specialagent/rule/method/MethodAgentRule.java

```diff
@@ -0,0 +1,74 @@
+package io.opentracing.contrib.specialagent.rule.method;
+
+import io.opentracing.contrib.specialagent.AgentRule;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.utility.JavaModule;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+/**
+ * MethodAgentRule
+ *
+ * @author code98@163.com
+ * @date 2019/11/15 5:27 下午
+ */
+public class MethodAgentRule extends AgentRule {
+
+    private static final String METHOD_CONFIG = "sa.config.instrumentation.plugin.method";
+
+    @Override
+    public Iterable<? extends AgentBuilder> buildAgent(AgentBuilder builder) throws Exception {
+        List<AgentBuilder> builders = new ArrayList<>();
+
+        String configInfos = System.getProperty(METHOD_CONFIG);
+        if (configInfos == null || "".equals(configInfos)) {
+            // sa.config.instrumentation.plugin.method is not config
+            return builders;
+        }
+
+        String[] classMethods = configInfos.split(";");
+        for (String classMethod : classMethods) {
+            String[] classMethodSplit = classMethod.split("#");
+            String className = classMethodSplit[0];
+            String method = classMethodSplit[1];
```

Comment:
@code98, please add error handling code. What if `classMethodSplit[1]` results in an `IndexOutOfBoundsException`?

Reply:
Hi @code98, I would like to propose a formal definition of your method instrumentation spec:

**Spec**
```
RULE := {CLASS}#{<init>({ARG1, ARG2,...})|{[~]METHOD}({ARG1, ARG2,...}):RETURN}:{<|>|^EXCEPTION}
RULES := {RULE}[;RULES]
```

**Usage**
`-Dsa.config.instrumentation.plugin.method={RULES}`

**Key**
`{...}`: Required.
`[...]`: Optional.
`CLASS`: Java identifier of a class name.
`<init>`: Identifies a constructor.
`METHOD`: Java identifier of a method name.
`ARG1, ARG2,...`: Java identifiers of argument class names.
`RETURN`: Java identifier of a class name.
`EXCEPTION`: Java identifier of an exception class name.
`~`: If included in front of a method name, the rule will match the method signature for all subclasses of `CLASS`.
`<`: If included at end of rule, the rule defines an intercept to be triggered on method exit.
`>`: If included at end of rule, the rule defines an intercept to be triggered on method enter.
`^EXCEPTION`: If included at end of rule, the rule defines an intercept to be triggered on exception `EXCEPTION`.
`;`: Rule delimiter.

This spec describes functionality that is more than what you have provided in this PR.
We can defer this expansion to the spec for a later update.

Reply:
Your rules are very detailed! But there is one more question that needs your answer.
`{<|> | ^ EXCEPTION}` is contradictory.
Why do we still need to distinguish between method enter and exit and exceptions? Because it is time-consuming information to calculate the specified method, it is necessary to enhance the code of the method enter and exit.


```
public static class MethodIntercept {

        @Advice.OnMethodEnter
        public static void enter(final @Advice.Origin String origin) {
            if (isEnabled(origin)) {
                MethodAgentIntercept.enter(origin);
            }
        }

        @Advice.OnMethodExit(onThrowable = Throwable.class)
        public static void exit(final @Advice.Origin String origin, final @Advice.Thrown Throwable t) {
            if (isEnabled(origin)) {
                MethodAgentIntercept.exit(t);
            }
        }
    }
```

Reply:
You can add two keys, `args` and `return`.
`args`: If this key exists, add the method's call parameters to the span, the value is json format.
`return`: If this key exists, add the return value of the method to the span, the value is json format.

Reply:
Hi @code98, please see my comments:

> Your rules are very detailed!

We'll need to document this plugin, so we can use this writeup in the README.md later.

> `{<|>|^EXCEPTION}` is contradictory.

Yes, the `|` characters mean "or".
It is necessary to provide a spec for ENTER, EXIT, and EXCEPTION, because there are situations where a method enters, runs for a very long time, and eventually exits -- or what if it is guaranteed to exit on exception? Your plugin applies to "all methods", which is fully generalized. This is why the plugin must consider general use-cases.

> We can also add two keys, `args` and `return`.

Can you provide an example?

Reply:
Sorry, I didn't understand the meaning of these three parameters.
I understand the | characters mean "or".

We have to trigger the rule on method enter and exit.

```
public static void enter(String origin) {

        // parse public java.lang.String com.test.HomeController.homePage(java.lang.String)

        String[] format1 = origin.split(" ");
        String methodNameInfo = format1[2];
        String[] format2 = methodNameInfo.split("\\(");
        String format3 = format2[0];
        String[] format4 = format3.split("\\.");
        String methodName = format4[format4.length - 1];

        final Span span = GlobalTracer.get()
                .buildSpan(methodName)
                .withTag(TAGS_KEY_SPAN_TYPE, TAGS_VALUE_INTERNAL)
                .withTag(TAGS_KEY_ORIGIN, origin)
                .withTag(Tags.COMPONENT.getKey(), "method").start();
        spanHolder.set(span);
    }

    public static void exit(Throwable t) {
        final Span span = spanHolder.get();
        if (span == null) {
            return;
        }
        if (t != null) {
            span.log(errorLogs(t));
            span.setTag(TAGS_KEY_ERROR, true);
            span.setTag(TAGS_KEY_ERROR_MESSAGE, t.getMessage());
            span.setTag(TAGS_KEY_HTTP_STATUS_CODE, TAGS_VALUE_500_INT);
        } else {
            span.setTag(TAGS_KEY_HTTP_STATUS_CODE, TAGS_VALUE_200_INT);
        }
        span.finish();
    }
```

Why make a distinction? Thank you for your answer

Reply:
Hi @code98, you're right. I was thinking about it from a different perspective. For the sake of lower complexity, let's leave it without the ENTER, EXIT, EXCEPTION spec.

Reply:
ok, from which viewpoint do you think about these three keys? Is it convenient to say in detail?

Reply:
Hi @safris , Examples of args and return parameters:

`args`
```
        @Advice.OnMethodEnter
        public static void enter(final @Advice.Origin String origin,
                                 final @Advice.AllArguments Object[] args) {
            
            if (rule contains args key) {
                 String argsJson = args to json string;
                 span.setTag("args", argsJson);
             }
            
        }
```

`return`
```
      @Advice.OnMethodExit(onThrowable = Throwable.class)
       public static void exit(final @Advice.Origin String origin,
                                final @Advice.Thrown Throwable t,
                                final @Advice.Return Object returnValue) {
            
            if (rule contains return key) {
                 String returnJson = returnValue to json string;
                 span.setTag("return", returnJson);
             }
        }
```

Reply:
Hi @code98, I had missed the point that you only cared to trace the span of a method ENTER to EXIT. Instead, I was assuming you wanted to create a span as a trigger, on ENTER, or on EXIT, or on EXCEPTION. I believe your use-case is more correct than mine.

Reply:
Regarding your `args` and `return`, could you specify how the JSON formal will look?

Reply:
Hi, @safris, 

java bean class:

![image](https://user-images.githubusercontent.com/10418235/70701786-e9a36e80-1d07-11ea-8f08-9e1c1a026f18.png)

Usage example:

![image](https://user-images.githubusercontent.com/10418235/70701899-23747500-1d08-11ea-8329-dd0e8726019f.png)


args json output:

```
[
    "value1",
    "value2"
]
```
return json output:

```
{
    "one":"1",
    "two":2
}
```


When a user defines a rule, it is necessary to ensure that this method must have a return value. Because the @ Advice.Return annotation document explains "If an instrumented method does not return a value, this annotation must not be used.
"

Reply:
Hi @code98, this sounds good. Please submit your changes to this PR and I will review it again.

FYI: Tomorrow I will release SpecialAgent v1.5.3. We can aim to release your plugin for v1.5.4, which will be on December 27th.

Reply:
Hi, @safris.  I think your location is in San Francisco. It's late, go to sleep! I now think about the development of the code.

Reply:
Hi, @safris  Rule example:  

RULE := {CLASS}#{<init>({ARG1,ARG2,...})|{[~]METHOD}({ARG1,ARG2,...}):RETURN}#[args,return]
RULES := {RULE}[;RULES]

1、Constructor with no parameters, no return value, No need for args and return
com.example.ClassA#<init>()#none

2、Constructor, 1 parameter, no return value, No need for args and return
com.example.ClassA#<init>(java.lang.String)#none

3、Normal method with no parameters, no return value, No need for args and return
com.example.ClassA#methodA():void#none

4、Normal method 2 parameters, The return value is com.example.Vo, No need for args and return
com.example.ClassA#methodB(java.lang.String,com.example.Model):com.example.Vo#none

5、All subclasses of ClassA, method name is methodC, 1 parameter, The return value is com.example.Vo, No need for args and return
com.example.ClassA#~methodC(java.lang.String):com.example.Vo#none

6、Ibid, Add args and return to the span
com.example.ClassA#~methodD(java.lang.String):com.example.Vo#args,return

Is there anything wrong?

Reply:
Hi @code98, thanks for the examples. These are a bit confusing though. 

1. If the method does not have parameters, the `()` can be removed. For example, the `Object.toString()` method is: `java.lang.Object#~toString:java.lang.String`.
1. A constructor's name is: `<init>`. For example, the default constructor for `Object` would be: `java.lang.Object#<init>`.
1. Another special intercept that's important to note is: class initialization (when the class is loaded and initialized by the class loader). This is referred to with `<clinit>`. For example: `java.net.URL#<clinit>`.
1. If there is no return, then just skip it (don't have to add `none`, and don't have to add `void`). What if there is a class in the `defaultPackage` that is called: `none`?! For example, the `Object#finalize()` method would be: `java.lang.Object#finalize`

I'm still confused as to how the `args` and `return` will work. I'll need to think about this some more.